File: modules/survey/dashboard/src/trail.ts
// Client-side trail extension mirrors the module's decimation (uiserve
// Trail) so points appended from live pose updates line up with what a
// later trail.get would return.
export const TRAIL_MIN_DIST_M = 0.15;
export const TRAIL_MIN_TURN_RAD = (10 * Math.PI) / 180;
export const TRAIL_CAP = 4000;

export type PosePoint = { x: number; y: number; theta: number };

export function wrapAngle(a: number): number {
  while (a > Math.PI) a -= 2 * Math.PI;
  while (a <= -Math.PI) a += 2 * Math.PI;
  return a;
}

export function shouldAppend(last: PosePoint | null, next: PosePoint): boolean {
  if (!last) return true;
  const moved = Math.hypot(next.x - last.x, next.y - last.y) >= TRAIL_MIN_DIST_M;
  const turned = Math.abs(wrapAngle(next.theta - last.theta)) >= TRAIL_MIN_TURN_RAD;
  return moved || turned;
}

export function appendCapped(pts: [number, number][], p: [number, number]): [number, number][] {
  const out = pts.length >= TRAIL_CAP ? pts.slice(pts.length - TRAIL_CAP + 1) : pts.slice();
  out.push(p);
  return out;
}

File: modules/survey/internal/uiserve/trail.go
package uiserve

import (
	"math"
	"sync"

	"github.com/waypointos/waypoint/modules/survey/internal/nav"
)

// Decimation thresholds; the dashboard panel mirrors them so locally
// extended points line up with what trail.get returns.
const (
	trailMinDistM   = 0.15
	trailMinTurnRad = 10 * math.Pi / 180
	trailCap        = 4000
)

// Trail keeps a decimated history of the believed pose for the map view.
type Trail struct {
	mu     sync.Mutex
	pts    [][2]float64
	last   nav.Pose
	primed bool
}

// Append records the pose if it moved or turned enough since the last kept
// point. Turn-only changes count too so in-place pivots stay visible.
func (t *Trail) Append(p nav.Pose) {
	t.mu.Lock()
	defer t.mu.Unlock()
	if t.primed {
		moved := math.Hypot(p.X-t.last.X, p.Y-t.last.Y) >= trailMinDistM
		turned := math.Abs(nav.WrapAngle(p.Theta-t.last.Theta)) >= trailMinTurnRad
		if !moved && !turned {
			return
		}
	}
	if len(t.pts) >= trailCap {
		t.pts = append(t.pts[:0], t.pts[1:]...)
	}
	t.pts = append(t.pts, [2]float64{p.X, p.Y})
	t.last, t.primed = p, true
}

func (t *Trail) Points() [][2]float64 {
	t.mu.Lock()
	defer t.mu.Unlock()
	out := make([][2]float64, len(t.pts))
	copy(out, t.pts)
	return out
}

func (t *Trail) Reset() {
	t.mu.Lock()
	defer t.mu.Unlock()
	t.pts, t.primed = nil, false
}
